autoscale/src/main/java/com/sequenceiq/periscope/monitor/evaluator/load/YarnLoadEvaluator.java

```diff
@@ -0,0 +1,180 @@
+
+    @Inject
+    private YarnMetricsClient yarnMetricsClient;
+
+    @Inject
+    private StackResponseUtils stackResponseUtils;
+
+    @Inject
+    private CloudbreakCommunicator cloudbreakCommunicator;
+
+    private long clusterId;
+
+    private Cluster cluster;
+
+    private LoadAlert loadAlert;
+
+    @Nonnull
+    @Override
+    public EvaluatorContext getContext() {
+        return new ClusterIdEvaluatorContext(clusterId);
+    }
+
+    @Override
+    public void setContext(EvaluatorContext context) {
+        clusterId = (long) context.getData();
+    }
+
+    @Override
+    public String getName() {
+        return EVALUATOR_NAME;
+    }
+
+    @Override
+    protected void execute() {
+        cluster = clusterService.findById(clusterId);
+        loadAlert = cluster.getLoadAlerts().stream().findFirst().get();
+
+        long remainingTime = ClusterUtils.getRemainingCooldownTime(
+                loadAlert.getLoadAlertConfiguration().getCoolDownMillis(),
+                cluster.getLastScalingActivity());
+
+        if (remainingTime <= 0) {
+            pollYarnMetricsAndScaleCluster();
+        } else {
+            LOGGER.debug("Cluster cannot be scaled for {} min(s)",
+                    ClusterUtils.TIME_FORMAT.format((double) remainingTime / TimeUtil.MIN_IN_MS));
+        }
+    }
+
+    protected void pollYarnMetricsAndScaleCluster() {
+        long start = System.currentTimeMillis();
+        try {
+            MDCBuilder.buildMdcContext(cluster);
+
+            StackV4Response stackV4Response = cloudbreakCommunicator.getByCrn(cluster.getStackCrn());
+
+            String hostGroupInstanceType =
+                    stackResponseUtils.getHostGroupInstanceType(stackV4Response, loadAlert.getScalingPolicy().getHostGroup());
+
+            YarnScalingServiceV1Response yarnResponse = yarnMetricsClient
+                    .getYarnMetricsForCluster(cluster, hostGroupInstanceType, stackV4Response.getCloudPlatform());
+
+            Map<String, String> hostFqdnsToInstanceId = stackResponseUtils
+                    .getCloudInstanceIdsForHostGroup(stackV4Response, loadAlert.getScalingPolicy().getHostGroup());
+            yarnResponse.getScaleUpCandidates().ifPresentOrElse(
+                    scaleUpCandidates -> handleScaleUp(hostGroupInstanceType, scaleUpCandidates, hostFqdnsToInstanceId.size()),
+                    () -> {
+                        yarnResponse.getScaleDownCandidates().ifPresent(
+                                scaleDownCandidates -> handleScaleDown(scaleDownCandidates, hostFqdnsToInstanceId));
+                    });
+
+            LOGGER.debug("Finished loadEvaluator for cluster {} in {} ms", cluster.getStackCrn(), System.currentTimeMillis() - start);
+        } catch (Exception ex) {
+            LOGGER.info("Failed to process load alert for Cluster {}, exception {}", cluster.getStackCrn(), ex);
+            eventPublisher.publishEvent(new UpdateFailedEvent(clusterId));
```

Comment:
Can more context be sent along with this event. What caused the failrue? (Inability to talk to YARN, something else)

Reply:
the exception is logged already which includes information like YARN unavailable or YARN 500 error, UpdateFailedEvent is used to handle other flow like tracking successive failures.

Reply:
eventPublished.publishEvent() ... is this what would eventually end up showing information on the 'Activities' page on the cluster UI?, or even to fetch scaling activity via the CLI?
If that is the case, it will be useful to have at least some information (Exception name, Exception message) to help with debugging without having to hunt through logs.

Reply:
ScaleUp\ScaleDown history on UI is not from periscope, that is shown from CB itself.